REGRESSION(263628@main): [GStreamer] Broke fast/mediastream/captureStream/canvas3d.html

philn opened the pull request:
#### b1285c6fa5649d15553718a936cb49c1c2ab12bc
<pre>
REGRESSION(263628@main): [GStreamer] Broke fast/mediastream/captureStream/canvas3d.html
<a href="https://bugs.webkit.org/show_bug.cgi?id=256758">https://bugs.webkit.org/show_bug.cgi?id=256758</a>

Reviewed by Xabier Rodriguez-Calvar.

Video frame conversions are now performed with the gst_video_convert_sample API which also takes
care of downloading textures from GPU when needed. The test almost passes now in WPE, still one
graphics-related failure to debug. In GTK the test still fails, again due to weak graphics support.

* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/wpe/fast/mediastream/captureStream/canvas3d-expected.txt: Added.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::videoFrameForCurrentTime):
* Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.cpp:
(WebCore::GStreamerVideoEncoder::encode):
(WebCore::GStreamerInternalVideoEncoder::encode):
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp:
(WebCore::convertSampleToImage):
(WebCore::VideoFrame::fromNativeImage):
(WebCore::VideoFrame::createNV12):
(WebCore::VideoFrame::createRGBA):
(WebCore::VideoFrame::createBGRA):
(WebCore::VideoFrame::createI420):
(WebCore::VideoFrameGStreamer::createFromPixelBuffer):
(WebCore::VideoFrame::copyTo):
(WebCore::VideoFrame::paintInContext):
(WebCore::VideoFrameGStreamer::resizedSample):
(WebCore::VideoFrameGStreamer::convert):
(WebCore::VideoFrameGStreamer::downloadSample):
(WebCore::VideoFrameGStreamer::convertToImage):
(WebCore::GstSampleColorConverter::singleton): Deleted.
(WebCore::GstSampleColorConverter::convertSample): Deleted.
(WebCore::GstSampleColorConverter::convertSampleToImage): Deleted.
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.h:

Canonical link: <a href="https://commits.webkit.org/265245@main">https://commits.webkit.org/265245@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/091e18ca90b8f034d769d21a286de76caa4fe1c2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10354 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10588 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10859 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11999 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9948 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10366 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12661 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10540 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12907 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10512 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/11292 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8725 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12395 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8595 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9379 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/16648 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/9689 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9530 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/12779 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/9970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/8093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/9129 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2486 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13380 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9821 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.cpp:
```diff
@@ -247,56 +247,14 @@ bool GStreamerInternalVideoEncoder::encode(VideoEncoder::RawFrame&& rawFrame, bo
-
-    return result;
+    auto convertedSample = gstVideoFrame.downloadSample();
+    return m_harness->pushSample(WTFMove(convertedSample));
```
You can do `m_harness->pushSample(gstVideoFrame.downloadSample())`;

philn committed:
REGRESSION(263628@main): [GStreamer] Broke fast/mediastream/captureStream/canvas3d.html
https://bugs.webkit.org/show_bug.cgi?id=256758

Reviewed by Xabier Rodriguez-Calvar.

Video frame conversions are now performed with the gst_video_convert_sample API which also takes
care of downloading textures from GPU when needed. The test almost passes now in WPE, still one
graphics-related failure to debug. In GTK the test still fails, again due to weak graphics support.

* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/wpe/fast/mediastream/captureStream/canvas3d-expected.txt: Added.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::videoFrameForCurrentTime):
* Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.cpp:
(WebCore::GStreamerVideoEncoder::encode):
(WebCore::GStreamerInternalVideoEncoder::encode):
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp:
(WebCore::convertSampleToImage):
(WebCore::VideoFrame::fromNativeImage):
(WebCore::VideoFrame::createNV12):
(WebCore::VideoFrame::createRGBA):
(WebCore::VideoFrame::createBGRA):
(WebCore::VideoFrame::createI420):
(WebCore::VideoFrameGStreamer::createFromPixelBuffer):
(WebCore::VideoFrame::copyTo):
(WebCore::VideoFrame::paintInContext):
(WebCore::VideoFrameGStreamer::resizedSample):
(WebCore::VideoFrameGStreamer::convert):
(WebCore::VideoFrameGStreamer::downloadSample):
(WebCore::VideoFrameGStreamer::convertToImage):
(WebCore::GstSampleColorConverter::singleton): Deleted.
(WebCore::GstSampleColorConverter::convertSample): Deleted.
(WebCore::GstSampleColorConverter::convertSampleToImage): Deleted.
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.h:

Canonical link: https://commits.webkit.org/265245@main
